File: cp.cpp
#include "signal_operation.hpp"
#include "updater.hpp"
#include "random_sampler.hpp"
#include "data.hpp"
#include "triple.hpp"
#include "model_file_writer.hpp"

#include <algorithm>
#include <iomanip>
#include <iostream>
#include <cmath>
#include <random>
#include <set>
#include <string>
#include <vector>

class CP {
public:
	SignalOperation op;
	Updater updater;
	std::vector< std::vector<double> > subjVec, objVec, relationVec;
	Data traindata;
	std::vector<double> outVec1, outVec2, outVec3;
	std::vector<double> norms;

	double fs;
	double lambda;
	int vecDim;
	std::mt19937 rnd; //int randomSeed;

	CP(Data& traindata, int vecDim_, int randomSeed, double learningRate, double threshold, double lambda_) : vecDim(vecDim_), lambda(lambda_) {
		op = SignalOperation();

		rnd = std::mt19937(randomSeed);
		double bnd = sqrt(6.0) / sqrt((double)vecDim * 2);
		std::uniform_real_distribution<> dist(-bnd, bnd);

		subjVec.resize(traindata.entity_counter);
		initializeModel(subjVec, traindata.entity_counter, dist);

		objVec.resize(traindata.entity_counter);
		initializeModel(objVec, traindata.entity_counter, dist);

		relationVec.resize(traindata.relation_counter);
		initializeModel(relationVec, traindata.relation_counter, dist);

		updater = Updater(learningRate, threshold);

		outVec1.resize(vecDim);
		outVec2.resize(vecDim);
		outVec3.resize(vecDim);

		norms.resize(3);
	}
	void initializeModel(std::vector< std::vector<double> >& Vec, const int counter, std::uniform_real_distribution<>& dist){
		for (int i = 0; i < counter; i++) {
			Vec[i].resize(vecDim);
			for (int j = 0; j < vecDim; j++) {
				Vec[i][j] = dist(rnd);
			}
		}
	}

	double sigmoid(double x) { return 1.0 / (1.0 + std::exp(-x));}

	double computeGradients(Triple& triple, double y) {
		const std::vector<double>& subj_v = subjVec[triple.subj];
		const std::vector<double>& obj_v = objVec[triple.obj];
		const std::vector<double>& relation_v = relationVec[triple.relation];

		op.product(subj_v, obj_v, relation_v, outVec1, outVec2, outVec3);

		double score = op.innerProduct(relation_v, outVec1);

		double yscore = y * score;
		double loss = std::log(1.0 + std::exp(-yscore));
		fs = -(y * sigmoid(-yscore));

		for (int i = 0; i < vecDim; i++) {
			outVec1[i] = fs * outVec1[i];
			outVec2[i] = fs * outVec2[i];
			outVec3[i] = fs * outVec3[i];
		}

		norms[0] = op.norm(outVec1);
		norms[1] = op.norm(outVec2);
		norms[2] = op.norm(outVec3);

		return loss;
	}

	void train(Data& data, int maxEpoch, int negativeSize, ModelFileWriter& mfw) {
		int sampleC = 1;
		double all = 0.0;
		double loss = 0.0;
		std::vector<Triple> train_triples = data.triples;
		std::set<long long> triple_key_set = data.triple_key_set;
		RandomSampler sampler(data.key_gen, data.entity_counter, data.relation_counter, rnd);

		for (int iter = 1; iter <= maxEpoch; ++iter) {
			std::shuffle(train_triples.begin(), train_triples.end(), rnd);
			std::cerr << "Epoch " << iter << ":" << std::endl;
			for (Triple& triple : train_triples) {
				if ((sampleC % 1000) == 0) {
					std::cerr << sampleC << " ";
					++all;
					updater.decay(all);
				}
				++sampleC;
				for (int i = 0; i <= negativeSize; i++) {
					if (i == 0) {
						loss += computeGradients(triple, 1.0);
						updater.SGD(triple, outVec1, outVec2, outVec3, norms, subjVec, objVec, relationVec);
					} else {
						Triple neg_triple = sampler.negativeSamplingSbj(triple, triple_key_set);
						loss += computeGradients(neg_triple, -1.0);
						updater.SGD(neg_triple, outVec1, outVec2, outVec3, norms, subjVec, objVec, relationVec);

						neg_triple = sampler.negativeSamplingObj(triple, triple_key_set);
						loss += computeGradients(neg_triple, -1.0);
						updater.SGD(neg_triple, outVec1, outVec2, outVec3, norms, subjVec, objVec, relationVec);
					}
				}
				if (updater.learningRate <= 0.0) {
					std::cerr << std::endl;
					std::cerr << "Learning rate was zero " << std::endl;
					std::cerr << "Total loss: " << loss << std::endl;
					mfw.save(subjVec, objVec, relationVec, op, data, iter);
				}
			}
			std::cerr << std::endl;
			std::cerr << std::fixed << std::setprecision(15) << "Total loss: " << loss << std::endl;

			loss = 0.0;
			sampleC = 1;

			if (iter % 10 == 0) {
				mfw.save(subjVec, objVec, relationVec, op, data, iter);
			}
		}
	}
};

int ArgPos(std::string str, int argc, char **argv) {
	int a;
	for (a = 1; a < argc; a++){
		if (str == argv[a]) {
			if (a == argc - 1) {
				std::cout << "Argument missing for" << str << std::endl;
				exit(1);
			}
			return a;
		}
	}
	return -1;
}

int main(int argc, char **argv){
	std::string trainname;
	int dimension;
	int iteration;
	double rate;
	double lambda;
	int i;
	if ((i = ArgPos((char *)"-train", argc, argv)) > 0) trainname = argv[i + 1];
	if ((i = ArgPos((char *)"-dimension", argc, argv)) > 0) dimension = atoi(argv[i + 1]);
	if ((i = ArgPos((char *)"-iteration", argc, argv)) > 0) iteration = atoi(argv[i + 1]);
	if ((i = ArgPos((char *)"-rate", argc, argv)) > 0) rate = std::stof(argv[i + 1]);
	if ((i = ArgPos((char *)"-lambda", argc, argv)) > 0) lambda = std::stof(argv[i + 1]);

	Data data;
	data.readFromRawFile(trainname);

	int randomSeed = 0;
	double threshold = 5.0;
	CP cp(data, dimension, randomSeed, rate, threshold, lambda);

	ModelFileWriter mfw;
	int negativeSize = 5;
	cp.train(data, iteration, negativeSize, mfw);

	return 0;
}

File: main/CP.java
package main;

import java.util.Collections;
import java.util.List;
import java.util.Random;
import java.util.Set;

public class CP {
	SignalOperation op;
	Updater updater;
	ParamVector[] subjVec, objVec, relationVec;
	TupleData traindata;
	double[] outVec1, outVec2, outVec3;
	double[] norms;
	
	double fs;
	double lambda;
	int vecDim; // vecDim_r;
	int randomSeed;
	
	public CP(
			TupleData traindata, 
			int vecDim, 
			int randomSeed,
			double learningRate, 
			double threshold,
			double lambda) 
	{
		this.traindata = traindata;
		op = new SignalOperation(/* vecDim */);
		subjVec = new ParamVector[traindata.entity_counter];
		objVec = new ParamVector[traindata.entity_counter];
		int i;
		for (i=0; i<traindata.entity_counter; ++i) {
			subjVec[i] = new ParamVector(vecDim);
			subjVec[i].initByUniformNormDist(i+randomSeed, op, false);
		}
		
		for (i=0; i<traindata.entity_counter; ++i) {
			objVec[i] = new ParamVector(vecDim);
			objVec[i].initByUniformNormDist(traindata.entity_counter+i+randomSeed, op, false);
		}
		relationVec = new ParamVector[traindata.relation_counter];
		for (i=0; i<traindata.relation_counter; ++i) {
			relationVec[i] = new ParamVector(vecDim);
			relationVec[i].initByUniformNormDist(traindata.entity_counter*2+i+randomSeed, op, false);
		}
		updater = new Updater(subjVec, objVec, relationVec, learningRate, threshold);
		outVec1 = new double[vecDim];
		outVec2 = new double[vecDim];
		outVec3 = new double[vecDim];
		this.vecDim = vecDim;
		this.lambda = lambda;
		norms = new double[3];
		this.randomSeed = randomSeed;
	}
	
	final public double sigmoid(double x) { return 1.d / (1.d + Math.exp(-x)); }
	
	public double computeGradients(Tuple tuple, double y) 
	{
		final double[] subj_v = subjVec[tuple.subj].values;
		final double[] obj_v = objVec[tuple.obj].values;
		final double[] relation_v = relationVec[tuple.relation].values;
		
		op.product(subj_v, obj_v, relation_v, outVec1, outVec2, outVec3);
		
		double score = op.innerProduct(relation_v, outVec1);
		
		double yscore = y * score;
		double loss = Math.log(1.d+Math.exp(-yscore));
		fs = -(y * sigmoid(-yscore));
		
		for (int i=0; i<vecDim; ++i) {
			outVec1[i] = fs * outVec1[i];
			outVec2[i] = fs * outVec2[i];
			outVec3[i] = fs * outVec3[i];
 		}
		
		norms[0] = op.norm(outVec1);
		norms[1] = op.norm(outVec2);
		norms[2] = op.norm(outVec3);
		
		if (loss == Double.POSITIVE_INFINITY || 
				loss == Double.NEGATIVE_INFINITY || 
				loss == Double.NaN) 
			System.err.println(yscore + " " + loss + " " + fs);
		return loss;
	}
	
	public void train(TupleData data, 
			RandomSampler sampler,
			int maxEpoch,
			int negativeSize,
			ModelFileWriter mfw) 
	{
		int sampleC = 1;
		double all = 0.d;
		double loss = 0.d;
		int i;
		List<Tuple> train_tuples = data.tuple_list;
		Set<Long> tuple_key_set = data.tuple_key_set;
		for (int iter = 1; iter <= maxEpoch; ++iter) {
			Collections.shuffle(train_tuples, new Random(iter+randomSeed));
			System.err.format("Epoch %d:\n", iter);
			for (Tuple tuple : train_tuples) {
				if ((sampleC % 1000) == 0) {
					System.err.format("%d ", sampleC);
					++all;
					updater.decay(all);
				}
				++sampleC;
				for (i=0; i<=negativeSize; ++i) {
					if (i == 0) {
						loss += computeGradients(tuple, 1.d);
						updater.SGD(tuple, outVec1, outVec2, outVec3, norms, lambda);
					} else {
						Tuple neg_tuple = sampler.negativeSamplingSbj(tuple, tuple_key_set);
						loss += computeGradients(neg_tuple, -1.d);
						updater.SGD(neg_tuple, outVec1, outVec2, outVec3, norms, lambda);
						
						neg_tuple = sampler.negativeSamplingObj(tuple, tuple_key_set);
						loss += computeGradients(neg_tuple, -1.d);
						updater.SGD(neg_tuple, outVec1, outVec2, outVec3, norms, lambda);
					}
				}
				if (updater.learningRate <= 0.d) {
					System.err.format("\n");
					System.err.format("Learning rate was zero \n");
					System.err.format("Total loss: %f\n", loss);
					if (mfw != null) {
						mfw.save(subjVec, objVec, relationVec, op, data, iter);
					}
					return ;
				}
			}
			System.err.format("\n");
			System.err.format("Total loss: %f\n", loss);
			loss = 0.d;
			sampleC = 1;
	
			if (mfw != null && iter % 10 == 0) {
				mfw.save(subjVec, objVec, relationVec, op, data, iter);
			}
		}
	}
}
